### ISSUE TITLE
Add Blob.uploadFrom(InputStream)

### ISSUE THREAD

**Issue by broady** [googleapis/java-storage#62]
We have [Blob.downloadTo(OutputStream)](https://googleapis.dev/java/google-cloud-clients/latest/com/google/cloud/storage/Blob.html#downloadTo-java.io.OutputStream-com.google.cloud.storage.Blob.BlobSourceOption...-) (and other downloadFrom variants)

Should probably have uploadFrom with a stream. People currently have to buffer/chunk their writes [manually](https://googleapis.dev/java/google-cloud-clients/latest/com/google/cloud/storage/Blob.html#writer-com.google.cloud.storage.Storage.BlobWriteOption...-).

/cc @crwilcox

**Comment by dmitry-fa** [comment]
Taking into account the fact that 
`Blob Storage.create(BlobInfo blobInfo, InputStream content, BlobWriteOption... options);`
is deprecated:
```
   * <p>This method is marked as {@link Deprecated} because it cannot safely retry, given that it
   * accepts an {@link InputStream} which can only be consumed once.
```
I don't think that it's a good idea to introduce:  `Blob.uploadFrom(InputStream)`
Instead I would suggest something like:

```
  public void uploadFrom(Path path, BlobWriteOption... options) {
    uploadFrom(path, 64*1024, options);
  }

  public void uploadFrom(Path path, int chunkSize, BlobWriteOption... options) {
    checkArgument(chunkSize > 0, "Position should be positive, is %d", chunkSize);
    try (WriteChannel writer = storage.writer(this, options)) {
      byte[] buffer = new byte[chunkSize];
      try (InputStream input = Files.newInputStream(path)) {
        int length;
        while ((length = input.read(buffer)) >= 0) {
          writer.write(ByteBuffer.wrap(buffer, 0, length));
        }
      }
    } catch (IOException e) {
      throw new StorageException(e);
    }
  }
```

**Comment by dmitry-fa** [comment]
New methods of Blob are suggested:

```
public void uploadFrom(Path path, BlobWriteOption... options) {...}
public void uploadFrom(Path path, int chunkSize, BlobWriteOption... options) {...}
public void uploadFrom(InputStream input, BlobWriteOption... options) {...}

  /**
   * Uploads the given content to this blob using specified blob write options and the provided
   * chunk size. The default chunk size is 2 MB. Larger chunk sizes may improve the upload
   * performance but require more memory. It could cause OutOfMemoryError or add significant garbage
   * collection overhead. Chunk sizes which are less than 256 KB are not allowed, they will be
   * treated as 256 KB.
   *
   * @param input content to be uploaded
   * @param chunkSize the minimum size that will be written by a single RPC.
   * @param options blob write options
   * @throws StorageException upon failure
   */
  public void uploadFrom(InputStream input, int chunkSize, BlobWriteOption... options) {
    chunkSize = Math.max(chunkSize, 262144); // adjust with MinChunkSize of BaseWriteChannel
    try (WriteChannel writer = storage.writer(this, options)) {
      writer.setChunkSize(chunkSize);
      byte[] buffer = new byte[chunkSize];
      int length;
      while ((length = input.read(buffer)) >= 0) {
        writer.write(ByteBuffer.wrap(buffer, 0, length));
      }
    } catch (IOException e) {
      throw new StorageException(e);
    }
  }
```

**Comment by dmitry-fa** [comment]
The following extension to the Blob class is proposed:

```
  public Blob uploadFrom(Path path, BlobWriteOption... options)
  public Blob uploadFrom(InputStream input, BlobWriteOption... options)

  /**
   * Uploads the given content to the storage using specified write channel and the given
   * buffer size. The default buffer size is 15 MB. Larger buffer sizes may improve the upload
   * performance but require more memory. It could cause OutOfMemoryError or add significant garbage
   * collection overhead. Buffer sizes which are less than 256 KB are not allowed, they will be
   * treated as 256 KB.
   *
   * <p>Note that this method does not close neither InputStream nor WriterChannel</p>
   *
   * <p>Example of uploading:
   *
   * <pre>{@code
   * Path file = Paths.get("humongous.file");
   * try (InputStream input = Files.newInputStream(file); WriteChannel writer = storage.writer(blobInfo)) {
   *     Blob.upload(input, writer, 150 * 1024 * 1024);
   * }
   *
   * @param input content to be uploaded
   * @param writer channel
   * @param bufferSize size of the buffer to read from input and send over writer
   */
  public static void upload(InputStream input, WriteChannel writer, int bufferSize)
```

**Comment by dmitry-fa** [comment]
Similar extension to the Storage interface is planned to be proposed with a separate PR

**Comment by dmitry-fa** [comment]
fix was reverted by #190

**Comment by dmitry-fa** [comment]
https://github.com/googleapis/java-storage/pull/214 fixes this issue as well
